docs(behaviorTests): Adding documentation regarding behavior unit tests (#693)
stardust-ui/react: .github/test-a-feature.md

yarn test:watch
```

[1]: https://github.com/stardust-ui/react/tree/master/test/specs/commonTests

> ## Behaviors tets

Reviewer: Behavior tests

Behavior unit tests are generated from the specification written in each behavior file.

> Each line under the "@specification" tag is taken and matched against the regex experesion written in "testDefinitions.ts" file. 

Reviewer 2: ```suggestion
Each line under the `@specification` tag is taken and matched against the regex expression written in `testDefinitions.ts` file. 
```

> **Adding test(s)**

Reviewer 2: ```suggestion
### Adding test(s)
```

> - For a new behavior. In the file 'behavior-test.tsx' do following changes:

Reviewer 3: For a new behavior, do following changes in the file 'behavior-test.tsx':

> - For a new behavior. In the file 'behavior-test.tsx' do following changes:

Reviewer 2: ```suggestion
- For a new behavior. In the file `behavior-test.tsx` do following changes:
```

  - import the new behavior into the test file

>   - add the new behavior to the testHelper object like: testHelper.addBehavior('yourNewBehaviorName', yourNewBehaviorImportedObject)

Reviewer 3: add the new behavior to the testHelper object like, for example:
```
 testHelper.addBehavior('yourNewBehaviorName', yourNewBehaviorImportedObject)
```

(use tripple backticks for the code block)

>   - add regex expression into 'testDefinitions.ts' which will match your line written under @specification tag

Reviewer 2: ```suggestion
  - add regex expression into `testDefinitions.ts` which will match your line written under `@specification` tag
```

- For an existing behavior:

>   - add regex expression into 'testDefinitions.ts' which will match your line written under @specification tag

Reviewer 2: ```suggestion
  - add regex expression into `testDefinitions.ts` which will match your line written under @specification tag
```

> **Running test(s)**

Reviewer 2: ```suggestion
### Running test(s)
```

Run test and watch: yarn jest --watch behavior-test

> **Troubleshooting**

Reviewer 2: ```suggestion
### Troubleshooting
```

> - I am not sure if my line under @specification was process correctly. <br>

Reviewer 3: remove `<br>`

> - I am not sure if my line under @specification was process correctly. <br>

Reviewer 2: ```suggestion
#### I am not sure if my line under `@specification` was process correctly.
```

Go into docs\src\behaviorMenu.json file and verify if you can find your line. If not then:
  - run command "gulp build:docs:component-menu-behaviors", this will build the file again

>   - verify formatting the file (if some tag is not missing, etc...) and run command again

Reviewer 3: verify formatting _of_ the file

> - I am not sure if my line was executed. </br> 

Reviewer 2: ```suggestion
#### I am not sure if my line was executed
```

> Rename all test files title containing "behavior-test" string. 

Reviewer 2: ```suggestion
Rename all test files title containing `behavior-test` string. 
```

  - listBehaviorrrrrrrrrrr-test.tsx
  - listItemBehaviorrrrrrr-test.tsx

Run the tests again and you should see in console:
```
 PASS  test/specs/behaviors/behavior-test.tsx
  buttonBehavior.ts
    √ Adds role='button' if element type is other than 'button'. (11ms)
    √ Adds attribute 'aria-disabled=true' based on the property 'disabled'. (8ms)
    √ Adds attribute 'aria-disabled=true' based on the property 'disabled'. This can be overriden by providing 'aria-disabled' property directly to the component.
  buttonGroupBehavior.ts
    √ Adds role 'presentation' to 'root' component's part (2ms)
    √ Wraps component in FocusZone allowing arrow key navigation through the children of the component.  
  dialogBehavior.ts
    √ Adds attribute 'aria-disabled=true' to 'trigger' component's part based on the property 'disabled'.
    √ Adds attribute 'aria-modal=true' to 'popup' component's part.
    √ Adds attribute 'role=dialog' to 'popup' component's part.
    √ Traps focus inside component
  gridBehavior.ts
    √ Wraps component in FocusZone allowing circular arrow key navigation through the children of the component.  
```

> - I want to add any description which should not be consider as unit test. <br>

Reviewer 3: remove `<br>`

> - I want to add any description which should not be consider as unit test. <br>

Reviewer 2: ```suggestion
#### I want to add any description which should not be consider as unit test
```

> Add description under the @description tag. Like:

Reviewer 2: ```suggestion
Add description under the `@description` tag. Like:
```

```
/**
 * @description
 * Image is usually only visual representation and therefore is hidden from screen readers.
```

> - I want to create unit tests in separate file not through the regex. <br>

Reviewer 3: remove `<br>`

> Add your spec file into the array of files "skipSpecChecksForFiles" in testHelper.tsx. And put description in behavior file under '@description" tag.

Reviewer 2: ```suggestion
Add your spec file into the array of files `skipSpecChecksForFiles` in `testHelper.tsx`. And put description in behavior file under `@description` tag.
```

yarn test:watch
```

[1]: https://github.com/stardust-ui/react/tree/master/test/specs/commonTests

> ## Behaviours tets

Reviewer: Behaviour tests

> Behavior unit tests are generated from the specification written in each behavior file.

Reviewer: For the sake of consistency, we can use either `behavior` - the American version, or `behaviour` the English version. And not both.

Reviewer 4: We used everywhere in namings `behavior` version. So, let's stick with this one

Author: thanks for catch going with behavior

> Each line under the "@specification" tag is taken and trying to match the regex experesion written in "testDefinitions.ts" file. 

Reviewer: maybe use `taken and matched against`?

**Adding test(s)**

> 

Reviewer: 1. For a new behavior:
- ...
- ...
2. For an existing behaviour:
- ...
- ...

What I need to do if I am adding ?:

> - New behavior file. In the file behavior-test.tsx add:

Reviewer: maybe specify where to add the file, also naming convention?

>   - your new added behavior into object: import {}

Reviewer: maybe use `import the new behavior into the test file` ?

>   - to testHelper like: testHelper.addBehavior('yourNewBehaviorName', yourNewBehaviorImportedObject)

Reviewer: maybe use `add it to the testHelper object like : testHelper.addBehavior('yourNewBehaviorName', yourNewBehaviorImportedObject)`

>   - add into testDefinitions.ts regex which will match your line written under @specification tag

Reviewer: for me, without context, this is not informative. add what? add the regex?

- Editing existing behavior
  - add into testDefinitions.ts regex which will match your line written under @specification tag

**Running test(s)**

Run test and watch: yarn jest --watch behavior-test

**Troubleshooting**

> - I am not sure if my line under @specification was process correctly? <br>

Reviewer: remove `?` since it's not a question

Go into docs\src\behaviorMenu.json file and verify if you can find your line. If not then:
  - run command "gulp build:docs:component-menu-behaviors", this will build the file again
  - verify formatting the file (if some tag is not missing, etc...) and run command again

> - I am not sure if my line was executed? </br> 

Reviewer: here as well, no `?`

```bash
yarn test:visual
```

## Behavior tets

Behavior unit tests are generated from the specification written in each behavior file.
Each line under the `@specification` tag is taken and matched against the regex expression written in `testDefinitions.ts` file.

### Adding test(s)

- For a new behavior. In the file `behavior-test.tsx` do following changes:
  - import the new behavior into the test file
  - add the new behavior to the testHelper object like: `testHelper.addBehavior('yourNewBehaviorName', yourNewBehaviorImportedObject)`
  - add regex expression into `testDefinitions.ts` which will match your line written under `@specification` tag

- For an existing behavior:

>   - add regex expression into `testDefinitions.ts` which will match your line written under @specification tag

Reviewer 2: ```suggestion
  - add regex expression into `testDefinitions.ts` which will match your line written under `@specification` tag
```

### Running test(s)

> Run test and watch: yarn jest --watch behavior-test

Reviewer 2: Run test and watch:

```sh
# yarn jest --watch behavior-test
```

### Troubleshooting

> #### I am not sure if my line under `@specification` was process correctly.

Reviewer 2: ```suggestion
#### I am not sure if my line under `@specification` was process correctly
```

> Go into docs\src\behaviorMenu.json file and verify if you can find your line. If not then:

Reviewer 2: ```suggestion
Go into `docs\src\behaviorMenu.json` file and verify if you can find your line. If not then:
```

>   - run command "gulp build:docs:component-menu-behaviors", this will build the file again

Reviewer 2: ```suggestion
  - run command `gulp build:docs:component-menu-behaviors`, this will build the file again
```

  - verify formatting the file (if some tag is not missing, etc...) and run command again

#### I am not sure if my line was executed
Rename all test files title containing `behavior-test` string.
For example, like (goal of the renaming is reach that these tests will not run):

>   - listBehaviorrrrrrrrrrr-test.tsx

Reviewer 2: ```suggestion
  - `listBehaviorrrrrrrrrrr-test.tsx`
```

>   - listItemBehaviorrrrrrr-test.tsx

Reviewer 2: ```suggestion
  - `listItemBehaviorrrrrrr-test.tsx`
```

Run the tests again and you should see in console:
```
 PASS  test/specs/behaviors/behavior-test.tsx
  buttonBehavior.ts
    √ Adds role='button' if element type is other than 'button'. (11ms)
    √ Adds attribute 'aria-disabled=true' based on the property 'disabled'. (8ms)
    √ Adds attribute 'aria-disabled=true' based on the property 'disabled'. This can be overriden by providing 'aria-disabled' property directly to the component.
  buttonGroupBehavior.ts
    √ Adds role 'presentation' to 'root' component's part (2ms)
    √ Wraps component in FocusZone allowing arrow key navigation through the children of the component.  
  dialogBehavior.ts
    √ Adds attribute 'aria-disabled=true' to 'trigger' component's part based on the property 'disabled'.
    √ Adds attribute 'aria-modal=true' to 'popup' component's part.
    √ Adds attribute 'role=dialog' to 'popup' component's part.
    √ Traps focus inside component
  gridBehavior.ts
    √ Wraps component in FocusZone allowing circular arrow key navigation through the children of the component.  
```

> #### I want to add any description which should not be consider as unit test.

Reviewer 2: ```suggestion
#### I want to add any description which should not be consider as unit test
```

Add description under the `@description` tag. Like:
```
/**
 * @description
 * Image is usually only visual representation and therefore is hidden from screen readers.
```

> #### I want to create unit tests in separate file not through the regex.

Reviewer 2: ```suggestion
#### I want to create unit tests in separate file not through the regex
```
